File: minimum-falling-path-sum.py
# Time Complexity : O(n^2) - n is len(matrix)
# Space Complexity : O(n^2) - m is len(matrix)
# Did this code successfully run on Leetcode : yes
# Any problem you faced while coding this : no

"""
This is similar to paint house, but instead of constant no. of colors 3, it is n here
Build a matrix to store the min sum for all possible paths
We can either pick 0th,1st,2nd or (n-1)th number from first row to get the min sum


To store the repeated sub problems, create matrix of n * n
Last row would be same as the input matrix's last row
From second last row, calculate and store the min sum of choosing element in that every column
And keep on doing that till we reach row 0
Min in row 0 will be the answer
"""

def minFallingPathSum(matrix):
    """
    :type matrix: List[List[int]]
    :rtype: int
    """
    n = len(matrix)

    # create matriz same as size as the input matrix
    dp = [[0 for i in range(0, n)] for i in range(0, n)]

    # copy the last row of matrix to dp matrix
    for i in range(0, n):
        dp[n - 1][i] = matrix[n - 1][i]

    # fill the matrix
    for i in range(n - 2, -1, -1):
        for j in range(0, n):
            # if first element, take min of that element and element after that
            if j == 0:
                dp[i][j] = matrix[i][j] + min(dp[i + 1][j], dp[i + 1][j + 1])
            # if last element, take min of that element and element before that
            elif j == n - 1:
                dp[i][j] = matrix[i][j] + min(dp[i + 1][j], dp[i + 1][j - 1])
            # if not corner elements, take min of that element, element after that and element before that
            else:
                dp[i][j] = matrix[i][j] + min(min(dp[i + 1][j - 1], dp[i + 1][j]), dp[i + 1][j + 1])
    return min(dp[0])

File: MinimumFallingPathSum.cs
// time complexity - O(n^2) where n= length of matrix
// space complexity - O(n) as we have used an 1-D array of length n
public class Solution
{
    public int MinFallingPathSum(int[][] matrix)
    {
        int n = matrix.Length;
        int[] dp = new int[n];

        for (int j = 0; j < n; j++)
        {
            dp[j] = matrix[n - 1][j];
        }
        for (int i = n - 2; i >= 0; i--)
        {
            int[] tmpDP = new int[n];
            for (int j = 0; j < n; j++)
            {
                // move straight down
                int down = dp[j];
                // move downLeft
                int downLeft = j > 0 ? dp[j - 1] : int.MaxValue;
                // move downright
                int downRight = j < n - 1 ? dp[j + 1] : int.MaxValue;
                // take min
                tmpDP[j] = matrix[i][j] + Math.Min(down, Math.Min(downLeft, downRight));
            }
            dp = tmpDP;
        }
        int minPathSum = int.MaxValue;
        for (int j = 0; j < n; j++)
        {
            minPathSum = Math.Min(minPathSum, dp[j]);
        }
        return minPathSum;
    }
}